homeassistant/components/tuneblade_remote/media_player.py

```diff
@@ -0,0 +1,161 @@
+        super().__init__(
+            coordinator,
+            coordinator.config_entry,
+            device_id,
+            device_data.get("name"),
+        )
+        self.device_id = device_id
+        self._attr_name = device_data["name"]
+        safe_name = self._attr_name.replace(" ", "_")
+        self._attr_unique_id = f"{device_id}@{safe_name}"
+        self._attr_volume_level = None
+        self._attr_state = MediaPlayerState.OFF
+        self._attr_supported_features = (
+            MediaPlayerEntityFeature.TURN_ON
+            | MediaPlayerEntityFeature.TURN_OFF
+            | MediaPlayerEntityFeature.VOLUME_SET
+        )
+
+    @property
+    def available(self) -> bool:
+        """Return True if the device is available."""
+        return self.device_id in self.coordinator.data
+
+    async def async_turn_on(self) -> None:
+        """Turn on the media player."""
+        await self.coordinator.client.connect(self.device_id)
+        await asyncio.sleep(5)
+        await self.coordinator.async_request_refresh()
+        self.hass.async_create_task(self._delayed_refresh())
+
+    async def async_turn_off(self) -> None:
+        """Turn off the media player."""
+        await self.coordinator.client.disconnect(self.device_id)
+        await asyncio.sleep(5)
+        await self.coordinator.async_request_refresh()
+        self.hass.async_create_task(self._delayed_refresh())
+
+    async def async_set_volume_level(self, volume: float) -> None:
+        """Set volume level for the media player."""
+        await self.coordinator.client.set_volume(self.device_id, int(volume * 100))
+        await self.coordinator.async_request_refresh()
+
+    async def _delayed_refresh(self) -> None:
+        await asyncio.sleep(5)
+        await self.coordinator.async_request_refresh()
+
+    async def async_added_to_hass(self) -> None:
+        """Register the coordinator update listener."""
+        await super().async_added_to_hass()
+        self.coordinator.async_add_listener(self._handle_coordinator_update)
+        self._handle_coordinator_update()
+
+    def _handle_coordinator_update(self) -> None:
+        assert self.device_id is not None
+        device_data = self.coordinator.data.get(self.device_id)
+        if device_data is None:
+            self._attr_state = MediaPlayerState.OFF
+            self._attr_volume_level = None
+        else:
+            code = str(device_data.get("status_code", "0"))
+            if code == "100":
+                self._attr_state = MediaPlayerState.PLAYING
+            elif code == "200":
+                self._attr_state = MediaPlayerState.IDLE
+            elif code == "0":
+                self._attr_state = MediaPlayerState.OFF
+            else:
+                self._attr_state = MediaPlayerState.OFF
+
+            volume = device_data.get("volume")
+            self._attr_volume_level = volume / 100 if volume is not None else None
+
+        self.async_write_ha_state()
```

Comment:
Split the logic to a separate function that you also call at the end of the constructor